File: 3646.next-special-palindrome-number.py
#
# @lc app=leetcode id=3646 lang=python3
#
# [3646] Next Special Palindrome Number
#

# @lc code=start
from itertools import permutations, combinations

class Solution:
    def specialPalindrome(self, n: int) -> int:
        s_n = str(n)
        len_n = len(s_n)
        
        # We iterate through lengths starting from the length of n
        # Since n <= 10^15, reasonable upper bound for length is enough.
        # Max possible sum of unique digits 1-9 is 45, but we stop as soon as we find a solution.
        for length in range(len_n, 46):
            candidates = []
            
            # Find all subsets of digits {1..9} that sum to 'length'
            # and satisfy parity constraints.
            # Parity constraint: 
            #   If length is even, all digits in subset must be even.
            #   If length is odd, exactly one digit in subset must be odd.
            
            valid_subsets = []
            
            # Helper to find subsets
            def find_subsets(target, current_digits, index):
                if target == 0:
                    valid_subsets.append(list(current_digits))
                    return
                if target < 0 or index > 9:
                    return
                
                # Include index
                find_subsets(target - index, current_digits + [index], index + 1)
                # Exclude index
                find_subsets(target, current_digits, index + 1)

            find_subsets(length, [], 1)
            
            for subset in valid_subsets:
                # Check parity constraints
                odds = [d for d in subset if d % 2 != 0]
                evens = [d for d in subset if d % 2 == 0]
                
                mid_digit = -1
                half_digits = []
                
                if length % 2 == 0:
                    if len(odds) != 0:
                        continue
                    # All evens, form half digits
                    for d in subset:
                        half_digits.extend([str(d)] * (d // 2))
                else:
                    if len(odds) != 1:
                        continue
                    mid_digit = odds[0]
                    # Add half counts for all numbers (odd one loses 1 for middle)
                    for d in subset:
                        half_digits.extend([str(d)] * (d // 2))
                
                # If we are strictly above the original length, we just want the smallest number
                # formed by these digits. Sort half_digits to get smallest prefix.
                if length > len_n:
                    half_digits.sort()
                    half_s = "".join(half_digits)
                    if length % 2 == 0:
                        cand_str = half_s + half_s[::-1]
                    else:
                        cand_str = half_s + str(mid_digit) + half_s[::-1]
                    candidates.append(int(cand_str))
                else:
                    # Length matches n. We need to check permutations to find strictly greater
                    # Since max length is around 16, half length is ~8. 8! is small (40k).
                    # We can generate unique permutations.
                    unique_perms = sorted(list(set(permutations(half_digits))))
                    for p in unique_perms:
                        half_s = "".join(p)
                        if length % 2 == 0:
                            cand_str = half_s + half_s[::-1]
                        else:
                            cand_str = half_s + str(mid_digit) + half_s[::-1]
                        
                        val = int(cand_str)
                        if val > n:
                            candidates.append(val)
            
            if candidates:
                return min(candidates)
                
        return -1 # Should not reach here

# @lc code=end

File: 3646.next-special-palindrome-number.cpp
#
# @lc app=leetcode id=3646 lang=cpp
#
# [3646] Next Special Palindrome Number
#
# @lc code=start
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    long long specialPalindrome(long long n) {
        static vector<long long> all;
        static bool inited = false;
        if (!inited) {
            inited = true;

            auto genForSubset = [&](const array<int,10>& cnt) {
                int len = 0;
                for (int d = 1; d <= 9; d++) len += cnt[d];
                if (len == 0 || len > 17) return;

                int odd = -1;
                for (int d = 1; d <= 9; d++) {
                    if (cnt[d] % 2 == 1) {
                        if (odd != -1) return; // more than one odd-count digit
                        odd = d;
                    }
                }
                if ((len % 2 == 0 && odd != -1) || (len % 2 == 1 && odd == -1)) return;

                int halfLen = len / 2;
                array<int,10> halfCnt{};
                for (int d = 1; d <= 9; d++) halfCnt[d] = cnt[d] / 2;

                string half(halfLen, '0');

                function<void(int)> dfs = [&](int pos) {
                    if (pos == halfLen) {
                        string s = half;
                        if (odd != -1) s.push_back(char('0' + odd));
                        string r = half;
                        reverse(r.begin(), r.end());
                        s += r;
                        all.push_back(stoll(s));
                        return;
                    }
                    for (int d = 1; d <= 9; d++) {
                        if (halfCnt[d] == 0) continue;
                        halfCnt[d]--;
                        half[pos] = char('0' + d);
                        dfs(pos + 1);
                        halfCnt[d]++;
                    }
                };

                dfs(0);
            };

            // enumerate all subsets of digits 1..9
            for (int mask = 1; mask < (1 << 9); mask++) {
                array<int,10> cnt{};
                int len = 0;
                for (int i = 0; i < 9; i++) {
                    if (mask & (1 << i)) {
                        int d = i + 1;
                        cnt[d] = d;
                        len += d;
                    }
                }
                if (len <= 17) genForSubset(cnt);
            }

            sort(all.begin(), all.end());
            all.erase(unique(all.begin(), all.end()), all.end());
        }

        auto it = upper_bound(all.begin(), all.end(), n);
        return *it; // guaranteed to exist for n <= 1e15
    }
};
# @lc code=end
